### DISCUSSION TITLE
Visualing surfaces/interfaces in 3D

### DISCUSSION THREAD

**Discussion by fu351** [mjhoptics/ray-optics#134] [Q&A] [answered]
I am currently trying to model the lens in 3D in my optical model system by tracing the rays and using the interception points(the coordinate when a ray enter/exits a surface/interface) to create a mesh to plot. 
However this has the issue of taking a long time when trying to create an accurate mesh since a lot of rays had to be traced to get enough interception points. 
Is there a way to model these surfaces directly by extracting their profiles instead of using the interception points?

**Comment by mjhoptics** [comment, marked as answer]
Hello @fu351,
Yes, I think you can do this in ray-optics without invoking ray-tracing.
The [Surface](https://ray-optics.readthedocs.io/en/latest/api/rayoptics.elem.surface.html#rayoptics.elem.surface.Surface) class has an instance of a [profile](https://ray-optics.readthedocs.io/en/latest/api/rayoptics.elem.profiles.html#module-rayoptics.elem.profiles) as an attribute. The [sag()](https://ray-optics.readthedocs.io/en/latest/api/rayoptics.elem.profiles.html#rayoptics.elem.profiles.SurfaceProfile.sag) method on the profile can be used to give the z coordinate of a point x, y on the surface.

```
N = 5      # mesh dimension
mesh_list = []
for i in range(1, sm.get_num_surfaces()-1):
    mesh = []
    surf = sm.ifcs[i]
    sd = surf.surface_od()
    for xj in np.linspace(-sd, sd, num=N):
        mesh_row = []
        for yk in np.linspace(-sd, sd, num=N):
            if surf.point_inside(xj, yk):
                z = surf.profile.sag(xj, yk)
            else:
                z = None
            mesh_row.append(z)
        mesh.append(mesh_row)
    mesh_list.append(mesh)
```
There's probably a more numpy-efficient way of doing this, but this conveys the idea.
Hope this helps.
Mike Hayford
